### PR TITLE
[ReaderFooter] update status bar in real time when inverting page turning

### PR DESCRIPTION
**Pull request by Commodore64user**
### what's new

* currently, the page-turn item introduced in #12249 will not automatically update the status bar. This removes the need to make manual changes (i.e turn the page) before changes are reflected on screen.
* update `onSwapPageTurnButtons` event handler to deal with all turn-page inversions
* adds two dispatcher entries so they can be used through Profiles.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12424)

### CODE REVIEW

**Comment by poire-z** [comment]
Probably - but that's for you to check if it's needed when you toggle it back to regular.

**Comment by NiLuJe** [comment]
Err, this isn't an Event that's sent by anything?

**Comment by NiLuJe** [comment]
By which I mean, the dispatcher event is `SwapPageTurnButtons`.

**Comment by Commodore64user** [comment]
> Probably - but that's for you to check if it's needed when you toggle it back to regular.

seems to work fine without anything else. As far as I can test.

> By which I mean, the dispatcher event appears to be SwapPageTurnButtons?
 
two of the other invertions don't have dispatcher events, the individual left/right page turn ones. I suppose that is what leads me to the question

**Comment by NiLuJe** [comment]
But `PageTurningInverted` is still not an actual event ;o).

```
┌─(niluje@ahsoka:pts/9)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~/Dev/koreader)─┐
└─(1:0.44:10%:00:45:%)── rg PageTurningInverted                                                                                                                                                                                                                                                                                                        ──(Thu, Aug 29)─┘
┌─(niluje@ahsoka:pts/9)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(~/Dev/koreader)─┐
└─(1:0.46:10%:00:53:%)──    
```

**Comment by Commodore64user** [comment]
> So this currently does a whole lot of nothing ^^.

through magic it is doing exactly what is supposed to... mmm am I not defining it just now, although somehow never calling it.

File: frontend/apps/reader/modules/readerfooter.lua [hunk 1]
```diff
@@ -2445,6 +2445,12 @@ function ReaderFooter:onNetworkConnected()
 end
 ReaderFooter.onNetworkDisconnected = ReaderFooter.onNetworkConnected
 
+function ReaderFooter:onPageTurningInverted()
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerfooter.lua:2448]:
```suggestion
function ReaderFooter: onSwapPageTurnButtons()
```

Comment (benoit-pierre) [reply frontend/apps/reader/modules/readerfooter.lua:2448]: Why? Is that coding style used elsewhere?

Comment (benoit-pierre) [reply frontend/apps/reader/modules/readerfooter.lua:2448]: Never mind, I was fixating on the stray space after the `:`.

**Comment by Commodore64user** [comment]
okay this is now throwing me off, because it works [somehow] on k4 but doesn't at all on voyage

**Comment by Commodore64user** [comment]
mmm, this is why k4 was automagically working

```lua
UIManager:scheduleIn(2, self.autoRefreshFooter)
```

**Comment by Commodore64user** [comment]
still not working, I am going to bed...

**Comment by Commodore64user** [comment]
This very quickly went from (falsely https://github.com/koreader/koreader/pull/12424#issuecomment-2316454188) being very promising to not working at all.

File: frontend/apps/reader/modules/readerfooter.lua [hunk 2]
```diff
@@ -2445,6 +2445,18 @@ function ReaderFooter:onNetworkConnected()
...
+        self:maybeUpdateFooter()
+    end
+end
+
+function ReaderFooter:onToggleReadingOrder()
+    if self.settings.page_turning_inverted then
+        self:maybeUpdateFooter()
+    end
+end
+
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readerfooter.lua:2459]:
```suggestion
ReaderFooter.onToggleReadingOrder = ReaderFooter.onSwapPageTurnButtons
```

(Instead of duplicating the code ;)).

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua:2459]: that's fine, but why doesn't it work? what am I missing?

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua:2459]: this works

```lua
function ReaderFooter:onSwapPageTurnButtons()
    UIManager:scheduleIn(0.1, self.autoRefreshFooter)
end
ReaderFooter.onToggleReadingOrder = ReaderFooter.onSwapPageTurnButtons
```
kind of, because for some reason it won't update if you toggle the swipes from the menu, swapping page-turn button from menu works though. with gestures both work fine. And the two page-turn inversion that don't have an event also don't work

File: frontend/apps/reader/modules/readerfooter.lua [hunk 3]
```diff
@@ -2446,16 +2446,9 @@ end
...
-    if self.settings.page_turning_inverted then
-        self:maybeUpdateFooter()
-    end
-end
-
-function ReaderFooter:onToggleReadingOrder()
-    if self.settings.page_turning_inverted then
-        self:maybeUpdateFooter()
-    end
+    UIManager:scheduleIn(0.1, self.autoRefreshFooter)
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readerfooter.lua:2449]: Nope.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua:2449]: Nope nope nope nope.

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua:2449]: one was enough but, what's the matter?

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua:2449]: This isn't how you do this, which tells me you're trying to fit a square peg into a round hole ;).

So: what is this trying to handle (i.e., where does that event emanate, and what's on screen at the time).

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua:2449]: > So: what is this trying to handle

when you use a gesture to trigger the page-turn button swap or the zone swap thing, update the status bar.

> and what's on screen at the time)?

the book one is reading.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua:2449]: Oh, ReaderFooter receives the event before DeviceListener, so the settings aren't actually swapped yet -_-".

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua]: > Oh, ReaderFooter receives the event before DeviceListener, so the settings aren't actually swapped yet -_-".

so that is why the maybeUpdateFooter() wouldn't work. Nice. I was losing my mind about it.

you also said, we should add it to callback for the menu ones, `UIManager:nextTick(xxx)` what would that xxx look like? 

https://github.com/koreader/koreader/blob/7e52c15de29c013413f3d9ffef37d2de86b44dbf/frontend/ui/elements/physical_buttons.lua#L48-L51

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua]: If the Menu is not actually in ReaderFooter, I'd rather said menu switch to events instead of reimplementing the same thing twice ;).

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua]: wait, are you saying: make those two individual page turn inversions an event on DeviceListener like onSwapPageTurnButtons?

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua]: No, I'm saying make the menu use the events instead of duplicating what the devicelistener handler does (I assume).

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua]: mmm, but that also doesn't update the status bar...

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua]: Something along the lines of

```diff
diff --git a/frontend/apps/reader/modules/readerfooter.lua b/frontend/apps/reader/modules/readerfooter.lua
index 255400087..7ec8d8ee4 100644
--- a/frontend/apps/reader/modules/readerfooter.lua
+++ b/frontend/apps/reader/modules/readerfooter.lua
@@ -2445,6 +2445,14 @@ function ReaderFooter:onNetworkConnected()
 end
 ReaderFooter.onNetworkDisconnected = ReaderFooter.onNetworkConnected
 
+function ReaderFooter:onSwapPageTurnButtons()
+    if self.settings.page_turning_inverted then
+        -- We may receive the event *before* DeviceListener, so delay this to make sure it had a chance to actually swap the settings.
+        UIManager:nextTick(self.maybeUpdateFooter, self)
+    end
+end
+ReaderFooter.onToggleReadingOrder = ReaderFooter.onSwapPageTurnButtons
+
 function ReaderFooter:onSetRotationMode()
     self:updateFooterContainer()
     self:resetLayout(true)
diff --git a/frontend/device/devicelistener.lua b/frontend/device/devicelistener.lua
index 8481f7670..456b47e94 100644
--- a/frontend/device/devicelistener.lua
+++ b/frontend/device/devicelistener.lua
@@ -310,15 +310,25 @@ function DeviceListener:onToggleFlashOnPagesWithImages()
     G_reader_settings:flipNilOrTrue("refresh_on_pages_with_images")
 end
 
-function DeviceListener:onSwapPageTurnButtons(show_notification)
+function DeviceListener:onSwapPageTurnButtons(show_notification, side)
     G_reader_settings:flipNilOrFalse("input_invert_page_turn_keys")
-    Device:invertButtons()
+    local key_name
+    if side == "left" then
+        Device:invertButtonsLeft()
+        key_name = _("Left-side page-turn buttons")
+    elseif side == "right" then
+        Device:invertButtonsRight()
+        key_name = _("Right-side page-turn buttons")
+    else
+        Device:invertButtons()
+        key_name = _("Page-turn buttons")
+    end
     if show_notification then
         local new_text
         if G_reader_settings:isTrue("input_invert_page_turn_keys") then
-            new_text = _("Page-turn buttons inverted.")
+            new_text = key_name .. " " .. _("inverted.")
         else
-            new_text = _("Page-turn buttons no longer inverted.")
+            new_text = key_name .. " " .. _("no longer inverted.")
         end
         Notification:notify(new_text)
     end
diff --git a/frontend/dispatcher.lua b/frontend/dispatcher.lua
index ea71d2f1d..ccba3c63e 100644
--- a/frontend/dispatcher.lua
+++ b/frontend/dispatcher.lua
@@ -80,7 +80,9 @@ local settingsList = {
     touch_input_off = {category="none", event="IgnoreTouchInput", arg=true, title=_("Disable touch input"), device=true},
     toggle_touch_input = {category="none", event="IgnoreTouchInput", title=_("Toggle touch input"), device=true, separator=true},
     ----
-    swap_page_turn_buttons = {category="none", event="SwapPageTurnButtons", arg=true, title=_("Invert page turn buttons"), device=true, condition=Device:hasKeys(), separator=true},
+    swap_left_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args={true, "left"}, title=_("Invert left-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
+    swap_right_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args={true, "right"}, title=_("Invert right-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
+    swap_page_turn_buttons = {category="none", event="SwapPageTurnButtons", arg=true, title=_("Invert page-turn buttons"), device=true, condition=Device:hasKeys(), separator=true},
     ----
     toggle_key_repeat = {category="none", event="ToggleKeyRepeat", title=_("Toggle key repeat"), device=true, condition=Device:hasKeys() and Device:canKeyRepeat(), separator=true},
     toggle_gsensor = {category="none", event="ToggleGSensor", title=_("Toggle accelerometer"), device=true, condition=Device:hasGSensor()},
@@ -306,6 +308,8 @@ local dispatcher_menu_order = {
     "touch_input_off",
     "toggle_touch_input",
     ----
+    "swap_left_page_turn_buttons",
+    "swap_right_page_turn_buttons",
     "swap_page_turn_buttons",
     ----
     "toggle_key_repeat",
diff --git a/frontend/ui/elements/physical_buttons.lua b/frontend/ui/elements/physical_buttons.lua
index 87c2a1e49..4f364d037 100644
--- a/frontend/ui/elements/physical_buttons.lua
+++ b/frontend/ui/elements/physical_buttons.lua
@@ -25,7 +25,7 @@ local PhysicalButtons = {
 
 if Device:hasDPad() and Device:useDPadAsActionKeys() then
     table.insert(PhysicalButtons.sub_item_table, {
-        text = _("Invert left-side page turn buttons"),
+        text = _("Invert left-side page-turn buttons"),
         enabled_func = function()
             return not G_reader_settings:isTrue("input_invert_page_turn_keys")
         end,
@@ -34,11 +34,11 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
         end,
         callback = function()
             G_reader_settings:flipNilOrFalse("input_invert_left_page_turn_keys")
-            Device:invertButtonsLeft()
+            UIManager:broadcastEvent(Event:new("SwapPageTurnButtons", false, "left"))
         end,
     })
     table.insert(PhysicalButtons.sub_item_table, {
-        text = _("Invert right-side page turn buttons"),
+        text = _("Invert right-side page-turn buttons"),
         enabled_func = function()
             return not G_reader_settings:isTrue("input_invert_page_turn_keys")
         end,
@@ -47,7 +47,7 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
         end,
         callback = function()
             G_reader_settings:flipNilOrFalse("input_invert_right_page_turn_keys")
-            Device:invertButtonsRight()
+            UIManager:broadcastEvent(Event:new("SwapPageTurnButtons", false, "right"))
         end,
         separator = true,
     })
```

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua]: Tweaked the delay above as is was screwing with Menu highlights.

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerfooter.lua]: `UIManager:scheduleIn(0.5, self.maybeUpdateFooter, self)`

nope nope nope nope. (sorry I couldn't resist... ;) )

I'll try it in a bit, will get back soon. the only thing I see from looking at it is, swapping one side [only] would now flip the G-setting for swapping both sides, is that intentional?

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua]: Oh, I may have forgotten to handle the settings properly, yeah ;D

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerfooter.lua]: Yup, definitely ;p.

File: frontend/device/devicelistener.lua [hunk 4]
```diff
@@ -311,14 +311,23 @@ function DeviceListener:onToggleFlashOnPagesWithImages()
 end
 
 function DeviceListener:onSwapPageTurnButtons(show_notification)
```
Comment (Commodore64user) [inline frontend/device/devicelistener.lua:313]:
```suggestion
function DeviceListener:onSwapPageTurnButtons(show_notification, side)
```

File: frontend/ui/elements/physical_buttons.lua [hunk 5]
```diff
@@ -8,14 +8,15 @@ local PhysicalButtons = {
...
-            text = _("Invert page turn buttons"),
+            text = _("Invert page-turn buttons"),
             enabled_func = function()
                 return not (G_reader_settings:isTrue("input_invert_left_page_turn_keys") or G_reader_settings:isTrue("input_invert_right_page_turn_keys"))
             end,
             checked_func = function()
                 return G_reader_settings:isTrue("input_invert_page_turn_keys")
             end,
             callback = function()
+                G_reader_settings:flipNilOrFalse("input_invert_page_turn_keys")
```
Comment (NiLuJe) [inline frontend/ui/elements/physical_buttons.lua:19]: No, that should definitely happen in the event handler ;).

**Comment by NiLuJe** [comment]
LGTM at a glance, letting others give it a once over ;).

File: frontend/device/devicelistener.lua [hunk 6]
```diff
@@ -310,15 +310,27 @@ function DeviceListener:onToggleFlashOnPagesWithImages()
...
+        key_name = _("Page-turn buttons")
+    end
     if show_notification then
         local new_text
         if G_reader_settings:isTrue("input_invert_page_turn_keys") then
-            new_text = _("Page-turn buttons inverted.")
+            new_text = key_name .. " " .. _("inverted.")
         else
-            new_text = _("Page-turn buttons no longer inverted.")
+            new_text = key_name .. " " .. _("no longer inverted.")
```
Comment (poire-z) [inline frontend/device/devicelistener.lua:333]: I'm not sure all this may play well with translations for some languages.

Comment (NiLuJe) [reply frontend/device/devicelistener.lua:333]: Huh, that's a fair point.

File: frontend/ui/elements/physical_buttons.lua [hunk 7]
```diff
@@ -25,29 +25,27 @@ local PhysicalButtons = {
 
 if Device:hasDPad() and Device:useDPadAsActionKeys() then
     table.insert(PhysicalButtons.sub_item_table, {
-        text = _("Invert left-side page turn buttons"),
+        text = _("Invert left-side page-turn buttons"),
```
[the remark this answers is not in the document]
Comment (NiLuJe) [reply frontend/ui/elements/physical_buttons.lua:28]: IIRC there was one where they only added a back button on the left, but generally, yeah, these are symmetric.

File: frontend/device/devicelistener.lua (hunk 6 above)
Comment (Commodore64user) [reply frontend/device/devicelistener.lua:333]: That can stay as it was before. Not a biggie.

Comment (Frenzie) [reply frontend/device/devicelistener.lua:333]: Almost certainly not. Being clever is rarely good in such matters.

**Comment by poire-z** [comment]
Looks ok, but not familiar with all that, trusting @NiLuJe .

**Review by Frenzie — APPROVED** [review]
lgtm but @NiLuJe knows this part of the code better

File: frontend/dispatcher.lua [hunk 8]
```diff
@@ -80,7 +80,9 @@ local settingsList = {
     touch_input_off = {category="none", event="IgnoreTouchInput", arg=true, title=_("Disable touch input"), device=true},
     toggle_touch_input = {category="none", event="IgnoreTouchInput", title=_("Toggle touch input"), device=true, separator=true},
     ----
-    swap_page_turn_buttons = {category="none", event="SwapPageTurnButtons", arg=true, title=_("Invert page turn buttons"), device=true, condition=Device:hasKeys(), separator=true},
+    swap_left_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args={true, "left"}, title=_("Invert left-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
```
Comment (hius07) [inline frontend/dispatcher.lua:83]: I doubt if using `args` is correct.
https://github.com/koreader/koreader/blob/491e24ea25e162b2f25e854232603eec986c7f56/frontend/dispatcher.lua#L1167-L1169
Does it work?

Comment (NiLuJe) [reply frontend/dispatcher.lua:83]: Huh, good point. I only checked that there were other actions specifying `args`, and assumed it was meant for the Event, but it doesn't look like that's quite what it's for (ConfigDialog/Option stuff?)

Comment (Commodore64user) [reply frontend/dispatcher.lua:83]: yes, I also only checked that it inverts the right side one, what I didn't notice was the it was in fact inverting both sets.

Comment (Commodore64user) [reply frontend/dispatcher.lua:83]: in this case, @NiLuJe given as how you didn't want to hard code the `show_notification` the other day, we could remove it and send only the side argument. you can fix the sources later

btw, thanks @hius07 for that catch.

**Comment by Commodore64user** [comment]
@NiLuJe, anything else that needs fixing here?

**Review by NiLuJe — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r5, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12424)* status: all files reviewed, 7 unresolved discussions (waiting on @Commodore64user, @hius07, and @poire-z)

___
*[`frontend/dispatcher.lua` line 83 at r9](https://reviewable.io/reviews/koreader/koreader/12424#-O689LqI9xl402zmEzYU:-O689LqI9xl402zmEzYV:b4rlltl) ([raw file](https://github.com/koreader/koreader/blob/c9ce3b5636bc70f802ebe62cab289d48ac8bdf68/frontend/dispatcher.lua#L83)):*
> ```Lua
>     toggle_touch_input = {category="none", event="IgnoreTouchInput", title=_("Toggle touch input"), device=true, separator=true},
>     ----
>     swap_left_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args="left", title=_("Invert left-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
> ```

Should be `arg`, not `args`
___
*[`frontend/dispatcher.lua` line 84 at r9](https://reviewable.io/reviews/koreader/koreader/12424#-O689SMG3MjrI_ke7kcB:-O689SMG3MjrI_ke7kcC:bs4q9op) ([raw file](https://github.com/koreader/koreader/blob/c9ce3b5636bc70f802ebe62cab289d48ac8bdf68/frontend/dispatcher.lua#L84)):*
> ```Lua
>     ----
>     swap_left_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args="left", title=_("Invert left-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
>     swap_right_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args="right", title=_("Invert right-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
> ```

Ditto.
___
*[`frontend/device/devicelistener.lua` line 326 at r9](https://reviewable.io/reviews/koreader/koreader/12424#-O689xmK1L8Mb1ArGxeI:-O689xmK1L8Mb1ArGxeJ:b-q0l24a) ([raw file](https://github.com/koreader/koreader/blob/c9ce3b5636bc70f802ebe62cab289d48ac8bdf68/frontend/device/devicelistener.lua#L326)):*
> ```Lua
>     local new_text
>     if G_reader_settings:isTrue("input_invert_page_turn_keys") or G_reader_settings:isTrue("input_invert_left_page_turn_keys")
>         or G_reader_settings:isTrue("input_invert_right_page_turn_keys") then
> ```

The exact message should probably be based on the arg, in case people do weird things.

Which means six full different strings, to deal with the original concerns about the concat ;).



_Code quote:_
```Lua
    if G_reader_settings:isTrue("input_invert_page_turn_keys") or G_reader_settings:isTrue("input_invert_left_page_turn_keys")
        or G_reader_settings:isTrue("input_invert_right_page_turn_keys") then
```

**Comment by Commodore64user** [comment]
would you prefer those `new_text` strings as `G_settings and x or y` or `if G_settings then x else y end`?

**Comment by NiLuJe** [comment]
> would you prefer those `new_text` strings as `G_settings and x or y` or `if G_settings then x else y end`?

Given the length of the keys and the strings, I think a full if then else construct would be more readable ;).

**Review by NiLuJe — APPROVED** [review]
> Reviewed 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12424)* status: all files reviewed, 3 unresolved discussions (waiting on @hius07 and @poire-z)

**Review by poire-z — APPROVED** [review]

**Comment by Commodore64user** [comment]
just noticed some [potential] shenanigans with dispatcher and custom shortcuts, updating now...

**Comment by poire-z** [comment]
The latest added message are clumsy. Isn't there a solution where you just don't change, or flip the other settings, without the "this is currently not possible, do other things first" ?

**Comment by Commodore64user** [comment]
It is unlikely that anyone will ever see this, but then again there are some 'very bad boys' out there. I also thought of a little dialogue that lets you choose to keep the current inversion or switch to the new one, but that seems like a lot more ugly (to me at least).

**Comment by poire-z** [comment]
Still, we dont really have anything like that (may be we have something for the PDF zoom/scroll combinations, some of which are probably invalid). And translators will see it.
`Currently, inverting the right-side page-turn buttons is not possible. Please revert the global page-turn inversion first.`
The user is using gestures/keymaps. He wants to do something, and you tell him: no, do toggle that blah blah blah - for which he doesn't care or even have no idea what the gesture/keymap to do that is.
Just do it for him, and if you want, tell him this was done (so, if he really cares, he can go find out how to revert the revert).

**Comment by Commodore64user** [comment]
> And translators will see it.

spooky. Seriously though, why is that a problem?

> Just do it for him, and if you want, tell him this was done (so, if he really cares, he can go find out how to revert the revert).

this is just about the worst way to handle this, as all the settings in the menu will be greyed out.

**Comment by Commodore64user** [comment]
how about something like 

```lua
elseif side == "right" then
        if G_reader_settings:isTrue("input_invert_page_turn_keys") then
            G_reader_settings:flipNilOrFalse("input_invert_page_turn_keys")
            Device:invertButtons()
        end
        G_reader_settings:flipNilOrFalse("input_invert_right_page_turn_keys")
        Device:invertButtonsRight()
        if G_reader_settings:isTrue("input_invert_right_page_turn_keys") then
            new_text = _("Right-side page-turn buttons inverted.")
        else
```
            new_text = _("Right-side page-turn buttons no longer inverted.")
        end

**Comment by poire-z** [comment]
That's indeed what I had in mind: if he sets "(only) right", disable the "left+right" without saying anything, feels logical.

**Comment by Frenzie** [comment]
What changed since I previously approved it?

**Comment by Commodore64user** [comment]
> What changed since I previously approved it?

you could invert an already inverted state and cause mayhem. it is better handled now. here onwards https://github.com/koreader/koreader/pull/12424#issuecomment-2334922459

File: frontend/device/devicelistener.lua [hunk 9]
```diff
@@ -310,18 +310,60 @@ function DeviceListener:onToggleFlashOnPagesWithImages()
...
 
-function DeviceListener:onSwapPageTurnButtons(show_notification)
-    G_reader_settings:flipNilOrFalse("input_invert_page_turn_keys")
-    Device:invertButtons()
-    if show_notification then
-        local new_text
+function DeviceListener:onSwapPageTurnButtons(side)
+    local new_text
+    if side == "left" then
+        -- revert any prior global inversions first. No shenanigans please, as we could end up with an all greyed out menu.
```
Comment (poire-z) [inline frontend/device/devicelistener.lua:316]: If there are ending periods, it should start like a sentence with an uppercase.
Also don't know what is conveyed by the wording "No shenanigans please", who are you addressing ? Reads a little odd.

**Review by poire-z — APPROVED** [review]

File: frontend/device/devicelistener.lua (hunk 9 above)
Comment (Commodore64user) [reply frontend/device/devicelistener.lua:331]: I am not sure how to explain it, it's like those signs you find in parks that read "No ball games" or "No dog fouling", they are aimed generally at the public, so not odd really

Comment (poire-z) [reply frontend/device/devicelistener.lua]: ~~Well, you should try to write it in plain english, for non-english future coders to not wonder the meaning.
For now, I don't know how to read it - and these are some alternative wording that I would know how to read, pick the one that is what you meant :)~~ {I stopped writting)
You edited it, thanks for that.

**Comment by Commodore64user** [comment]
Is anything else missing here?

**Review by NiLuJe — APPROVED** [review]
> Reviewed 1 of 2 files at r13, 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12424)* status: all files reviewed, 4 unresolved discussions (waiting on @Commodore64user, @Frenzie, @hius07, and @poire-z)

**Review by poire-z — APPROVED** [review]

**Comment by mergen3107** [comment]
@Commodore64user 
Congrats on merging!

Can you please explain how this PR works?

Is it going to invert direction of progress bar (the one that fills in as book progresses) when RTL page turning enabled?

If so I can test this with manga, where I have directory defaults enabling RTL page turns based on folder name.

**Comment by NiLuJe** [comment]
Nope, it's about the "reading direction" status bar element (which may be somewhat new, as someone not relying on RTL, I wasn't actually aware of it before testing this ;)).

**Comment by Commodore64user** [comment]
> @Commodore64user Congrats on merging!

thanks!

> Can you please explain how this PR works?

sure. It is a lot less exciting than the title makes it sound though.

> Is it going to invert direction of progress bar (the one that fills in as book progresses) when RTL page turning enabled?

Nope, i mean, I honestly didn’t even know that was a thing until one duckduckgo search ago https://developer.apple.com/design/human-interface-guidelines/right-to-left. But it does sound like a very good idea to implement.

this pr only makes sure that the status bar item (the two arrows) are displayed immediately after triggering the inversion (buttons or swipes) via a gesture, shortcut or from the menu (technically speaking, half a second after). That’s all really. For reference, previously you would either have to manually change the page or wait for something else to update the status bar (the time for example) before it would show up there.

there are also some more bits related to page turn buttons on NT devices, but that is not relevant for your case, i guess.

**Comment by mergen3107** [comment]
Thanks guys!

This will really help me in my current investigation though :D 

I will create a proper issue ticket later when I have more time. 

In short, some of my manga books (marked as finished, I think) open up without directory defaults for some reason. I will see when exactly this happens.

**Comment by Commodore64user** [comment]
> (which may be somewhat new

yes, fresh out of the oven #12249. Although I originally only intended to use it for physical page turn inversions, @poire-z strongly suggested ;) we included swipe/tap inversions as well
